action/protocol/staking/handler_stake_migrate.go

```diff
@@ -0,0 +1,208 @@
+		return &handleError{
+			err:           errors.New("cannot migrate unstaked bucket"),
+			failureStatus: iotextypes.ReceiptStatus_ErrInvalidBucketType,
+		}
+	}
+	if err := validateBucketWithoutEndorsement(NewEndorsementStateManager(csm.SM()), bucket, protocol.MustGetBlockCtx(ctx).BlockHeight); err != nil {
+		return err
+	}
+	return validateBucketSelfStake(protocol.MustGetFeatureCtx(ctx), csm, bucket, false)
+}
+
+func (p *Protocol) withdrawBucket(ctx context.Context, withdrawer *state.Account, bucket *VoteBucket, cand *Candidate, csm CandidateStateManager) (*receiptLog, *action.TransactionLog, error) {
+	// delete bucket and bucket index
+	if err := csm.delBucketAndIndex(bucket.Owner, bucket.Candidate, bucket.Index); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to delete bucket for candidate %s", bucket.Candidate.String())
+	}
+
+	// update bucket pool
+	if err := csm.CreditBucketPool(bucket.StakedAmount); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to update staking bucket pool %s", err.Error())
+	}
+	// update candidate vote
+	weightedVote := p.calculateVoteWeight(bucket, false)
+	if err := cand.SubVote(weightedVote); err != nil {
+		return nil, nil, &handleError{
+			err:           errors.Wrapf(err, "failed to subtract vote for candidate %s", bucket.Candidate.String()),
+			failureStatus: iotextypes.ReceiptStatus_ErrNotEnoughBalance,
+		}
+	}
+	// clear candidate's self stake if the
+	if cand.SelfStakeBucketIdx == bucket.Index {
+		cand.SelfStake = big.NewInt(0)
+		cand.SelfStakeBucketIdx = candidateNoSelfStakeBucketIndex
+	}
+	if err := csm.Upsert(cand); err != nil {
+		return nil, nil, csmErrorToHandleError(cand.GetIdentifier().String(), err)
+	}
+	// update withdrawer balance
+	if err := withdrawer.AddBalance(bucket.StakedAmount); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to add balance %s", bucket.StakedAmount)
+	}
+	// put updated withdrawer's account state to trie
+	actionCtx := protocol.MustGetActionCtx(ctx)
+	if err := accountutil.StoreAccount(csm.SM(), actionCtx.Caller, withdrawer); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to store account %s", actionCtx.Caller.String())
+	}
+	// create receipt log
+	actLog := newReceiptLog(p.addr.String(), HandleWithdrawStake, protocol.MustGetFeatureCtx(ctx).NewStakingReceiptFormat)
+	actLog.AddTopics(byteutil.Uint64ToBytesBigEndian(bucket.Index), bucket.Candidate.Bytes())
+	actLog.AddAddress(actionCtx.Caller)
+	actLog.SetData(bucket.StakedAmount.Bytes())
+	return actLog, &action.TransactionLog{
+		Type:      iotextypes.TransactionLogType_WITHDRAW_BUCKET,
+		Amount:    bucket.StakedAmount,
+		Sender:    address.StakingBucketPoolAddr,
+		Recipient: actionCtx.Caller.String(),
+	}, nil
+}
+
+func (p *Protocol) ConstructExecution(ctx context.Context, act *action.MigrateStake, sr protocol.StateReader) (*action.Execution, error) {
+	sm := protocol.NewStateManagerWrapper(sr)
+	csm, err := p.constructCandidateStateManager(ctx, sm)
```

Comment:
i think all you need is:
```
csr, err := ConstructBaseView(sr)
```
so we can avoid the `constructCandidateStateManager` and `NewStateManagerWrapper`

Reply:
and with this change:
```
func (p *Protocol) fetchBucket(csm BucketGetByIndex, index uint64) (*VoteBucket, ReceiptError) {
```